### DISCUSSION TITLE
[Immunization] How to calculate DUE dates for dependent Tasks | Code Changes

### DISCUSSION THREAD

**Discussion by dubdabasoduba** [opensrp/fhircore#2140] [Ideas]
# Context

1. https://github.com/opensrp/fhircore/discussions/2141
2. https://github.com/opensrp/fhircore/discussions/2142

# Implementation

## How to identify the link between the questionnaire items and a Tasks

- **Individual tasks form**
    - The current FHIR Core implementation already links a Questionnaire and a Task through
      the `Task.reasonReference.reference` attribute.
    - This is then used to update the `Task.status` after the Questionnaire is filled and
      the `QuestionnaireResponse.status` marked as `completed`.
    - The function to update the DUE dates should be called immediately after updating the `Task.status` attribute.
- **Record all tasks form**
    - This form will be used to record administration dates for multiple tasks.
    - The tasks are represented as `Questionnaire.item.answerOption` on the first question as
      described [here](https://github.com/opensrp/fhircore/discussions/2139).
        - The answer options should have the following characteristics.
            - The `Questionnaire.item.answerOption.valueCoding.id` as the `Task.id`.
            - The `Questionnaire.item.answerOption.valueCoding.display` as the `Task.description`.
            - The `Questionnaire.item.answerOption.valueCoding.code` as the `Task.id`.
    - Based on the option we go with [here](https://github.com/opensrp/fhircore/discussions/2143), we have 2 options on
      how to generate the inputs for the update DUE dates function below.
        1. **Tasks completion via StructureMaps**
            1. The extracted resources are added as contained resources on the `QuestionnaireResponse.contained`
               attribute. Iterate through the `QuestionnaireResponse.contained` to get the Tasks then use the Tasks list
               to generate the input for the update DUE function below.
        2. **Tasks completion via code**
            1. Write a function to extract the `Task.id` from the resulting `QuestionnaireResponse.item.answer` array
               and fetch the tasks related to the fetched ids from the local app database.
            2. Use the resulting tasks list to generate the input for the update DUE date function below.

## How to calculate DUE dates for dependent Tasks

- Write a function that runs immediately after a Task is completed.
- This function will update the date on the `Task.excecutionPeriod.start` attribute to account for the prerequisite
  administration date.
    - This will affect the time the Task becomes DUE.
    - The method should only execute if
        - The dependent task has the prerequisite referenced on the `Task.partOf` attribute.
        - If the dependent task is DUE, UPCOMING or OVERDUE.
        - The dependent task has an `Task.input.type` attribute defined with
          the `{"coding":[{"system":"http://snomed.info/sct","code":"371154000","display":"Dependent (qualifier value)"}]}`.
        - The Task should also have the prerequisite tasks referenced on the `Task.input` attribute with the reference
          code on the `Task.input.type` attribute.
            - Here is an
              example `{"type":{"coding":[{"system":"http://snomed.info/sct","code":"900000000000457003","display":"Reference set attribute (foundation metadata concept)"}]},"reference":"Task/<string-uuid>"}`.
- The new `Task.excutionPeriod.start` date will be calculated as follows
    - Get the administration date added to the Immunization recording QuestionnaireResponse.
        - Let's call it `administration_date`. This can be found on the immunization resource here https://github.com/opensrp/fhircore/discussions/2142
    - Subtract the `administration_date` from the dependent task `Task.excutionPeriod.start` to get the number of days
      between the 2 dates.
        - Let's call the difference `date_interval`.
    - Check if the `date_interval` is `=>` to the time interval _(e.g. 28)_ defined on the dependent
      task `Task.input.value` [here](https://github.com/opensrp/fhircore/discussions/2141).
        - If yes, do nothing. otherwise, add the time interval provided on the dependent task `Task.input.value` to
          the `administration_date` and update the dependent task `Task.excutionPeriod.start` to the results.

**Comment by pld** [comment]
- "The worker should only run if" isn't it that the worker always runs, but the below specifies what Tasks the worker searches for to update?
- Are these 2 bullets saying the same/an equivalent thing?
>    - The dependent task has an input.type attribute defined with
the {"coding":[{"system":"http://snomed.info/sct","code":"900000000000457003","display":"Reference set attribute (foundation metadata concept)"}]}.
>    - The Task should also have the prerequisite tasks referenced on the input attribute with the reference code
on the input.type attribute.
>        - Here is an example {"type":{"coding":[{"system":"http://snomed.info/sct","code":"900000000000457003","display":"Reference set attribute (foundation metadata concept)"}]},"reference":"Task/<string-uuid>"}.
- How is the "Immunization recording form" linked to the Task?
- Can we make explicit somewhere that the value in `Task.input.value` is in days?

**Reply by dubdabasoduba** [reply]
Updated.
